### ISSUE TITLE
Flexible loss configuration for nested/multi outputs losses learning in Keras

### ISSUE THREAD

**Issue by nicolaspi** [keras-team/keras#20351]
Following the [guidelines](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md#step-1-open-an-issue).
### Description

Currently, Keras has limited support for specifying losses for complex, nested model outputs. I propose an enhancement to allow more flexible loss configuration using a dictionary that maps output paths to loss functions. This approach would greatly improve the usability of Keras for models with intricate output structures.

### Proposed Solution

The key idea is to allow users to specify losses using string paths that represent entries in the output structure. These paths would be separated by dots (`.`) (by default) and could refer to either leaf nodes or subtrees in the output structure.

Key features of this proposal:

1. Paths can represent nested dictionary keys or list indices (e.g., `'a.b.c'` or `'a.b.1'`).
2. Paths don't need to represent leaf nodes; they can refer to subtrees, allowing for more granular loss application.
3. The entire subtree at a given path would be passed to the corresponding loss function.

### Usage Example

```python
model = keras.Model(inputs=inputs, outputs={'a': y1, 'b': {'c': (y1, y2), 'd': y3 }})

loss = {'a': loss_fn, 'b': loss_fn, 'b.c': loss_fn, 'b.d': loss_fn }

model.compile(optimizer='adam', loss=loss, jit_compile=True)
```

In this example, different loss functions are applied to different parts of the output structure.

I made a [POC](https://github.com/nicolaspi/keras-nested-loss/blob/master/main.py) demonstrating the approach:

```bash
5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - a_loss: 1.1311 - b.c_loss: 1.1311 - b.d_loss: 1.1519 - b_loss: 1.1311 - loss: 4.5453
Epoch 2/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - a_loss: 0.9445 - b.c_loss: 0.9445 - b.d_loss: 1.0794 - b_loss: 0.9445 - loss: 3.9128
Epoch 3/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - a_loss: 0.7808 - b.c_loss: 0.7808 - b.d_loss: 1.0327 - b_loss: 0.7808 - loss: 3.3750
Epoch 4/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - a_loss: 0.6338 - b.c_loss: 0.6338 - b.d_loss: 0.9995 - b_loss: 0.6338 - loss: 2.9008
```

**Comment by nicolaspi** [comment]
Closing, PR: https://github.com/keras-team/keras/pull/20358
